### PR TITLE
Engine vault layer & Transaction send refactor OK-6440 OK-6801 OK-6597 OK-6492 OK-6486 OK-6537 OK-6433 OK-7070 OK-7069 OK-7120 OK-7122  OK-6770 

### PR DESCRIPTION
**Pull request by sidmorizon**
wiki:
https://onekeyhq.atlassian.net/wiki/spaces/ONEKEY/pages/128778245/Sign+and+Send+Transaction

### CODE REVIEW

File: packages/components/src/Modal/Container/useModalClose.ts [hunk 1]
```diff
@@ -15,6 +16,8 @@ function useModalClose({ onClose }: { onClose?: () => void | boolean }) {
     if (parent?.canGoBack()) {
       parent?.goBack();
```
Comment (sidmorizon) [inline packages/components/src/Modal/Container/useModalClose.ts:16]: 如果 parent 不存在，再判断 navigation 是否可回退

File: packages/engine/src/index.ts [hunk 2]
```diff
@@ -98,20 +100,33 @@ import {
...
 
   readonly validator: Validators;
 
+  public vaultFactory = new VaultFactory({
+    engine: this,
+  });
+
   constructor() {
+    // throw new OneKeyInternalError();
+    // throw new OneKeyInternalError('custom internal error message');
```
Comment (sidmorizon) [inline packages/engine/src/index.ts]: 删除

File: packages/engine/src/index.ts [hunk 3]
```diff
@@ -1063,6 +1080,79 @@ class Engine {
...
+    encodedTx: any;
+    password: string;
+    networkId: string;
+    accountId: string;
+  }) {
+    const vault = await this.getVault({
+      accountId,
+      networkId,
+    });
+    console.log('signAndSendTransaction >>>> encodedTx', encodedTx);
```
Comment (sidmorizon) [inline packages/engine/src/index.ts]: 改成 debugLogger

File: packages/engine/src/index.ts [hunk 4]
```diff
@@ -1063,6 +1080,79 @@ class Engine {
...
+    accountId: string;
+    encodedTx: any;
+  }) {
+    const vault = await this.vaultFactory.getVault({ networkId, accountId });
+
+    return vault.fetchFeeInfo(encodedTx);
+  }
+
+  @backgroundMethod()
+  async appendFeeInfoToEncodedTx(params: {
```
Comment (sidmorizon) [inline packages/engine/src/index.ts]: attachFeeInfo

File: packages/engine/src/index.ts [hunk 5]
```diff
@@ -1063,6 +1080,79 @@ class Engine {
...
+    encodedTx: IEncodedTxAny;
+    feeInfoValue: IFeeInfoUnit;
+  }): Promise<IEncodedTxAny> {
+    const { networkId, accountId } = params;
+    const vault = await this.vaultFactory.getVault({ networkId, accountId });
+    return vault.appendFeeInfoToEncodedTx(params);
+  }
+
+  @backgroundMethod()
+  async buildEncodedTxFromTransfer({
```
Comment (sidmorizon) [inline packages/engine/src/index.ts:1129]: fromTransfer
fromTransferNft
fromSwap

File: packages/engine/src/index.ts [hunk 6]
```diff
@@ -1063,6 +1080,79 @@ class Engine {
...
+    networkId: string;
+    accountId: string;
+    transferInfo: ITransferInfo;
+  }) {
+    const vault = await this.vaultFactory.getVault({ networkId, accountId });
+    return vault.buildEncodedTxFromTransfer(transferInfo);
+  }
+
   @backgroundMethod()
   async prepareTransfer(
```
Comment (sidmorizon) [inline packages/engine/src/index.ts:1067]: 删除

File: packages/engine/src/index.ts [hunk 7]
```diff
@@ -1123,18 +1213,59 @@ class Engine {
...
+      value,
+      gasPrice,
+      gasLimit,
+      tokenIdOnNetwork,
+      extra,
+    );
+  }
+
   @backgroundMethod()
   async transfer(
```
Comment (sidmorizon) [inline packages/engine/src/index.ts:1127]: 删除

File: packages/engine/src/proxy.ts [hunk 8]
```diff
@@ -122,7 +123,88 @@ function fromDBNetworkToChainInfo(dbNetwork: DBNetwork): ChainInfo {
...
+    ],
+    type,
+    nonce,
+    feeLimit,
+    feePricePerUnit: feePricePerUnit?.shiftedBy(network.feeDecimals),
+    payload,
+  };
+}
+
+export function fillUnsignedTx(
```
Comment (sidmorizon) [inline packages/engine/src/proxy.ts:207]: 删除

File: packages/engine/src/proxy.ts [hunk 9]
```diff
@@ -122,7 +123,88 @@ function fromDBNetworkToChainInfo(dbNetwork: DBNetwork): ChainInfo {
   };
 }
 
-function fillUnsignedTx(
+export function fillUnsignedTx2({
```
Comment (sidmorizon) [inline packages/engine/src/proxy.ts]: 挪 vaults 层

File: packages/engine/src/proxy.ts [hunk 10]
```diff
@@ -591,10 +674,14 @@ class ProviderController extends BaseProviderController {
       }
     } else {
       const result = await this.getFeePricePerUnit(networkId);
+      // const fast = {
```
Comment (sidmorizon) [inline packages/engine/src/proxy.ts]: 删除

File: packages/ext/src/ui/resizeEventOptimize.ts [hunk 11]
```diff
@@ -17,6 +43,9 @@ function optimize() {
       options: any,
       ...others: any
     ) => {
+      // if (!['click', 'error'].includes(type)) {
+      //   console.log('resizeEventOptimize addEventListener >>>>>> ', { type });
```
Comment (sidmorizon) [inline packages/ext/src/ui/resizeEventOptimize.ts]: 删除

File: packages/kit/src/views/Send/Authentication.tsx [hunk 12]
```diff
@@ -30,12 +30,25 @@ const SendAuth: FC<EnableLocalAuthenticationProps> = ({
...
   password,
 }) => {
-  const navigation = useNavigation();
+  const navigation = useNavigation<NavigationProps>();
   const toast = useToast();
   const intl = useIntl();
+  const route = useRoute<RouteProps>();
   const { dispatch } = backgroundApiProxy;
 
   const createTransfer = useCallback(async () => {
```
Comment (sidmorizon) [inline packages/kit/src/views/Send/Authentication.tsx:38]: rename sendTx

File: packages/kit/src/views/Send/Send.tsx [hunk 13]
```diff
@@ -47,26 +51,61 @@ type TransactionValues = {
 
 type Option = SelectItem<string>;
 
+// function GeneralSignModal({ encodedTx }) {
```
Comment (sidmorizon) [inline packages/kit/src/views/Send/Send.tsx]: 删除

File: packages/kit/src/views/Send/Send.tsx [hunk 14]
```diff
@@ -47,26 +51,61 @@ type TransactionValues = {
...
+//   // encodedTx -> decodedTx -> decodedPlainTx
+//   const decodedPlainTx = encodedTx;
+//   const contentView = null;
+//   if (decodedPlainTx.type === 'transfer') {
+//     contentView = null;
+//   }
+// }
+
+const buildEncodedTxFromTransferDebounced = debounce(
```
Comment (sidmorizon) [inline packages/kit/src/views/Send/Send.tsx:55]: createPromisedDebounce

File: packages/kit/src/views/Send/Send.tsx [hunk 15]
```diff
@@ -47,26 +51,61 @@ type TransactionValues = {
...
+
 const Transaction = () => {
+  // const encodedTxRef = useRef<any>(null);
+  const [encodedTx, setEncodedTx] = useState(null);
   const navigation = useNavigation<NavigationProps>();
   const { control, handleSubmit, watch } = useForm<TransactionValues>({
     mode: 'onSubmit',
     reValidateMode: 'onBlur',
+    defaultValues: {
+      to: '0xa9b4d559a98ff47c83b74522b7986146538cd4df',
```
Comment (sidmorizon) [inline packages/kit/src/views/Send/Send.tsx]: remove

File: packages/kit/src/views/Send/Send.tsx [hunk 16]
```diff
@@ -336,11 +414,19 @@ const Transaction = () => {
                 </Typography.Body2Strong>
 
                 <Pressable
+                  disabled={!feeInfoPayload && !feeInfoPayloadLoading}
```
Comment (sidmorizon) [inline packages/kit/src/views/Send/Send.tsx]: or

File: packages/kit/src/views/Send/Send.tsx [hunk 17]
```diff
@@ -47,26 +51,61 @@ type TransactionValues = {
...
+const buildEncodedTxFromTransferDebounced = debounce(
+  (options: {
+    networkId: string;
+    accountId: string;
+    transferInfo: ITransferInfo;
+    callback: (promise: Promise<any>) => void;
+  }) => {
+    const { callback, ...others } = options;
+    console.log('calling buildEncodedTxFromTransferDebounced', others);
+    callback(backgroundApiProxy.engine.buildEncodedTxFromTransfer(others));
```
Comment (sidmorizon) [inline packages/kit/src/views/Send/Send.tsx:63]: try catch return null

File: packages/kit/src/views/Send/Send.tsx [hunk 18]
```diff
@@ -100,89 +139,127 @@ const Transaction = () => {
...
+  // build transferInfo
+  // TODO move to watch useEffect below
+  const transferInfo: ITransferInfo = useDeepCompareMemo(() => {
+    // TODO token undefined
+    const { to, value } = formFields;
+    const from = (account as { address: string }).address;
+    const info = {
+      from,
+      to,
+      token: selectedToken?.tokenIdOnNetwork,
```
Comment (sidmorizon) [inline packages/kit/src/views/Send/Send.tsx:149]: tokenId
buildEncodedTxFromTransferInfo

File: packages/kit/src/views/Send/Send.tsx [hunk 19]
```diff
@@ -100,89 +139,127 @@ const Transaction = () => {
...
-      to: data.to,
-      account: {
-        id: account.id,
-        name: account.name,
-        address: (account as { address: string }).address,
-      },
-      network: {
-        id: activeNetwork?.network.id ?? '',
-        name: activeNetwork?.network.name ?? '',
+      payload: {
```
Comment (sidmorizon) [inline packages/kit/src/views/Send/Send.tsx:209]: transferPayload
transferNftPayload
swapPayload

sendPayloadTransfer

File: packages/kit/src/views/Send/Send.tsx [hunk 20]
```diff
@@ -100,89 +139,127 @@ const Transaction = () => {
...
     const params = {
-      to: data.to,
-      account: {
-        id: account.id,
-        name: account.name,
-        address: (account as { address: string }).address,
-      },
-      network: {
-        id: activeNetwork?.network.id ?? '',
-        name: activeNetwork?.network.name ?? '',
```
Comment (sidmorizon) [inline packages/kit/src/views/Send/Send.tsx:149]: payloadType=transfer

File: packages/kit/src/views/Send/Send.tsx [hunk 21]
```diff
@@ -336,11 +414,19 @@ const Transaction = () => {
                 </Typography.Body2Strong>
 
                 <Pressable
```
Comment (sidmorizon) [inline packages/kit/src/views/Send/Send.tsx:338]: standAloneComponent

File: packages/kit/src/views/Send/SendEditFee.tsx [hunk 22]
```diff
@@ -190,20 +298,128 @@ const EditFeeTabs = ({
...
+    let priceInfo: string | EIP1559Fee = data.gasPrice;
+    if (isEIP1559Fee) {
+      priceInfo = {
+        baseFee: data.baseFee,
+        maxPriorityFeePerGas: data.maxPriorityFeePerGas,
+        maxFeePerGas: data.maxFeePerGas,
+      };
+    }
+    navigation.navigate({
+      name: SendRoutes.Send,
```
Comment (sidmorizon) [inline packages/kit/src/views/Send/SendEditFee.tsx:352]: custom back route name

**Review by ByteZhang1024 — DISMISSED** [review]

**Review by ByteZhang1024 — APPROVED** [review]

**Review by linleiqin — APPROVED** [review]

File: package.json [hunk 23]
```diff
@@ -112,6 +111,7 @@
     "webpack-cli": "3.3.12",
     "webpack-dev-server": "^4.3.1",
     "worker-loader": "^3.0.7",
-    "axios": "^0.26.0"
+    "axios": "^0.26.0",
+    "concurrently": "^7.0.0"
```
Comment (revan-zhang) [inline package.json:115]: 这两个没有按照字母表顺序排列好奇怪..

**Review by revan-zhang — APPROVED** [review]

**Pull request merged**